compression_verifier.go

```diff
@@ -0,0 +1,224 @@
+package ghostferry
+
+import (
+	"crypto/md5"
+	"database/sql"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"strings"
+
+	sq "github.com/Masterminds/squirrel"
+	"github.com/golang/snappy"
+	"github.com/siddontang/go-mysql/schema"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	// CompressionSnappy is used to identify Snappy (https://google.github.io/snappy/) compressed column data
+	CompressionSnappy = "SNAPPY"
+)
+
+// UnsupportedCompressionError is used to identify a
+type UnsupportedCompressionError struct {
+	table     string
+	column    string
+	algorithm string
+}
+
+func (e UnsupportedCompressionError) Error() string {
+	return "Compression algorithm: " + e.algorithm +
+		" not supported on table: " + e.table +
+		" for column: " + e.column
+}
+
+// CompressionVerifier provides the same functionality as the iterative verifier but also
+// provides support for manually verifying the payload of compressed columns that may
+// have different hashes for the same data
+type CompressionVerifier struct {
+	logger *logrus.Entry
+
+	// supportedAlgorithms provide O(1) lookup to check if a configured algorithm is supported
+	supportedAlgorithms map[string]struct{}
+
+	// tableCmpressions is represented as table[column][compression_algorithm]
+	tableCompressions map[string]map[string]string
+}
+
+// NewCompressionVerifier first checks the map for supported compression algorithms before
+// initializing and returning the initialized instance.
+func NewCompressionVerifier(tableCompressions map[string]map[string]string) (*CompressionVerifier, error) {
+	supportedAlgorithms := make(map[string]struct{})
```

Comment:
nitpick: Could `supportedAlgorithms` maybe be a constant instead of a variable on every single `CompressionVerifier` "instance"? After all this can't possibly change (without recompiling the binary).

Reply:
Go doesn't allow map constants, but we could scope this at the package level and put it in `init` (although it wouldn't be exported so it would be local to this file). I usually try to avoid that because it [adds a bit of magic and is bad](https://peter.bourgon.org/blog/2017/06/09/theory-of-modern-go.html), though I did consider it. Keeping it in the constructor does scope it to the instantiated `CompressionVerifier`, and only one instance should ever be needed.